File: src/minecraft/net/machinemuse/api/electricity/UEElectricAdapter.java
package net.machinemuse.api.electricity;

import net.machinemuse.powersuits.common.ModCompatability;
import net.minecraft.item.ItemStack;
import universalelectricity.core.electricity.ElectricityPack;
import universalelectricity.core.item.IItemElectric;

public class UEElectricAdapter extends ElectricAdapter {
	protected ItemStack stack;
	protected IItemElectric item;

	public UEElectricAdapter(ItemStack stack) {
		this.stack = stack;
		this.item = (IItemElectric) stack.getItem();
	}

	@Override
	public double getCurrentEnergy() {
		double currentEnergy = museEnergyFromJoules(item.getJoules(stack));
		return currentEnergy;
	}

	@Override
	public double getMaxEnergy() {
		double maxEnergy = museEnergyFromJoules(item.getMaxJoules(stack));
		return maxEnergy;
	}

	@Override
	public double drainEnergy(double requested) {
		double voltage = item.getVoltage(stack);
		ElectricityPack requestedPack = museEnergyToElectricityPack(requested, voltage);
		ElectricityPack receivedPack = item.onProvide(requestedPack, stack);
		return museEnergyFromElectricityPack(receivedPack);
	}

	@Override
	public double giveEnergy(double provided) {
		double voltage = item.getVoltage(stack);
		ElectricityPack packToProvide = museEnergyToElectricityPack(provided, voltage);
		ElectricityPack eatenPack = item.onReceive(packToProvide, stack);
		return museEnergyFromElectricityPack(eatenPack);
	}

	public static ElectricityPack museEnergyToElectricityPack(double museEnergy, double voltage) {
		double coulombsPerTick = museEnergyToJoules(museEnergy);
		double amps = coulombsPerTick * 20.0 / voltage;
		return new ElectricityPack(amps, voltage);
	}

	public static double museEnergyFromElectricityPack(ElectricityPack pack) {
		double coulombsPerTick = pack.amperes / 20.0 * pack.voltage;
		double museEnergy = museEnergyFromJoules(coulombsPerTick);
		return museEnergy;
	}

	public static double museEnergyToJoules(double energy) {
		return energy / ModCompatability.getUERatio();
	}

	public static double museEnergyFromJoules(double energy) {
		return energy * ModCompatability.getUERatio();
	}

}

File: src/minecraft/net/machinemuse/api/electricity/UEElectricAdapter.scala
package net.machinemuse.api.electricity

import net.machinemuse.powersuits.common.ModCompatability
import net.minecraft.item.ItemStack
import universalelectricity.core.electricity.ElectricityPack
import universalelectricity.core.item.IItemElectric
import net.machinemuse.api.electricity.PowerConversions._

class UEElectricAdapter(val stack:ItemStack) extends ElectricAdapter {
  val item = stack.getItem.asInstanceOf[IItemElectric]


  def getCurrentEnergy: Double = {
    val currentEnergy: Double = museEnergyFromJoules(item.getJoules(stack))
    return currentEnergy
  }

  def getMaxEnergy: Double = {
    val maxEnergy: Double = museEnergyFromJoules(item.getMaxJoules(stack))
    return maxEnergy
  }

  def drainEnergy(requested: Double): Double = {
    val voltage: Double = item.getVoltage(stack)
    val requestedPack: ElectricityPack = museEnergyToElectricityPack(requested, voltage)
    val receivedPack: ElectricityPack = item.onProvide(requestedPack, stack)
    return museEnergyFromElectricityPack(receivedPack)
  }

  def giveEnergy(provided: Double): Double = {
    val voltage: Double = item.getVoltage(stack)
    val packToProvide: ElectricityPack = museEnergyToElectricityPack(provided, voltage)
    val eatenPack: ElectricityPack = item.onReceive(packToProvide, stack)
    return museEnergyFromElectricityPack(eatenPack)
  }
}